File: python/src/anjo_core/engine.py
"""Async end-to-end orchestration over injected model, store, and retriever adapters."""

from __future__ import annotations

import re
from collections.abc import Callable, Iterable, Sequence
from dataclasses import dataclass, replace
from datetime import UTC, datetime
from typing import Literal

from .affect import TurnShapePolicy, decoding_params, turn_shape_directive
from .appraisal import AppraisalPolicyInput, AppraisalResult, default_appraisal_policy
from .models import (
    CognitionState,
    CompanionState,
    GateInput,
    GateResult,
    GenerateInput,
    MemoryCandidate,
    Message,
    PresenceVector,
    RetrievalInput,
    TurnResult,
)
from .prompt import PromptInputs, PromptPolicy, build_system_prompt, build_untrusted_context
from .protocols import AppraisalPolicy, MemoryRetriever, ModelAdapter, StateStore
from .retrieval import rank_candidates
from .surfacing import build_presence_vector

DEFAULT_GATE_RESULT = GateResult(intent="CASUAL", should_respond=True, should_retrieve=False)
SILENT_GATE_RESULT = GateResult(intent="CASUAL", should_respond=False, should_retrieve=False)
BUILTIN_INTENTS = frozenset(
    {"ABUSE", "APOLOGY", "VULNERABILITY", "CURIOSITY", "CHALLENGE", "NEGLECT", "CASUAL"}
)
GateErrorMode = Literal["raise", "respond", "silent"]
_INTENT_PATTERN = re.compile(r"^[A-Z][A-Z0-9_]{0,63}$")


@dataclass(frozen=True, slots=True)
class EngineLimits:
    """Positive resource ceilings applied by :class:`CompanionEngine`."""

    max_message_chars: int = 16_000
    max_history_messages: int = 200
    max_history_chars: int = 128_000
    max_prompt_chars: int = 128_000
    max_output_chars: int = 32_000

    def __post_init__(self) -> None:
        for name in (
            "max_message_chars",
            "max_history_messages",
            "max_history_chars",
            "max_prompt_chars",
            "max_output_chars",
        ):
            value = getattr(self, name)
            if isinstance(value, bool) or not isinstance(value, int):
                raise TypeError(f"{name} must be an integer")
            if value <= 0:
                raise ValueError(f"{name} must be positive")


DEFAULT_ENGINE_LIMITS = EngineLimits()


def normalize_intent(intent: str, *, custom_intents: Iterable[str] = ()) -> str:
    """Normalize a gate label and reject unregistered custom behavior.

    Built-in labels drive deterministic appraisal. Registered custom labels are
    passed through in normalized form and intentionally receive no built-in
    appraisal impulse.
    """
    if not isinstance(intent, str):
        raise TypeError("gate intent must be a string")
    normalized = intent.strip().upper()
    allowed_custom = frozenset(_normalize_custom_intent(item) for item in custom_intents)
    if normalized not in BUILTIN_INTENTS and normalized not in allowed_custom:
        raise ValueError(f"unsupported gate intent: {intent!r}")
    return normalized


def _normalize_custom_intent(intent: str) -> str:
    if not isinstance(intent, str):
        raise TypeError("custom intents must be strings")
    normalized = intent.strip().upper()
    if not _INTENT_PATTERN.fullmatch(normalized):
        raise ValueError("custom intents must match [A-Z][A-Z0-9_]{0,63}")
    return normalized


class CompanionEngine:
    """Run gate → retrieve → appraise → affect → generate with explicit seams.

    Gate and generation are model-driven and are not parity claims. The state
    transforms between them are deterministic and credential-free.
    """

    def __init__(
        self,
        *,
        model: ModelAdapter,
        store: StateStore,
        conversation_id: str,
        retriever: MemoryRetriever | None = None,
        base_prompt: str = "",
        prompt_policy: PromptPolicy | None = None,
        turn_shape_policy: TurnShapePolicy | None = None,
        appraisal_policy: AppraisalPolicy = default_appraisal_policy,
        state_factory: Callable[[], CompanionState] = CompanionState,
        retrieval_limit: int = 6,
        max_retrieval_candidates: int = 64,
        gate_error_mode: GateErrorMode = "raise",
        custom_intents: Iterable[str] = (),
        limits: EngineLimits = DEFAULT_ENGINE_LIMITS,
    ) -> None:
        if not conversation_id:
            raise ValueError("conversation_id must not be empty")
        if isinstance(retrieval_limit, bool) or not isinstance(retrieval_limit, int):
            raise TypeError("retrieval_limit must be an integer")
        if retrieval_limit < 0:
            raise ValueError("retrieval_limit must be non-negative")
        if isinstance(max_retrieval_candidates, bool) or not isinstance(
            max_retrieval_candidates, int
        ):
            raise TypeError("max_retrieval_candidates must be an integer")
        if max_retrieval_candidates <= 0:
            raise ValueError("max_retrieval_candidates must be positive")
        if retrieval_limit > max_retrieval_candidates:
            raise ValueError("retrieval_limit must not exceed max_retrieval_candidates")
        if gate_error_mode not in ("raise", "respond", "silent"):
            raise ValueError("gate_error_mode must be 'raise', 'respond', or 'silent'")
        if not isinstance(limits, EngineLimits):
            raise TypeError("limits must be EngineLimits")
        self.model = model
        self.store = store
        self.retriever = retriever
        self.conversation_id = conversation_id
        self.base_prompt = base_prompt
        self.prompt_policy = prompt_policy or PromptPolicy()
        self.turn_shape_policy = turn_shape_policy or TurnShapePolicy()
        self.appraisal_policy = appraisal_policy
        self.state_factory = state_factory
        self.retrieval_limit = retrieval_limit
        self.max_retrieval_candidates = max_retrieval_candidates
        self.gate_error_mode = gate_error_mode
        self.custom_intents = frozenset(_normalize_custom_intent(item) for item in custom_intents)
        self.limits = limits

    async def turn(
        self,
        text: str,
        *,
        on_token: Callable[[str], None] | None = None,
    ) -> TurnResult:
        """Process and atomically persist one store-serialized conversation turn."""
        if not isinstance(text, str):
            raise TypeError("text must be a string")
        if len(text) > self.limits.max_message_chars:
            raise ValueError(f"message exceeds {self.limits.max_message_chars} characters")
        async with self.store.transaction(self.conversation_id) as transaction:
            history = await transaction.load_transcript()
            self._validate_history(history)
            state = await transaction.load_state() or self.state_factory()
            gate = await self._gate(text, history, state)
            user_message = Message("user", text)
            if not gate.should_respond:
                await transaction.commit(state=None, messages=(user_message,))
                return TurnResult(
                    text="",
                    intent=gate.intent,
                    silent=True,
                    mood=state.mood,
                )

            now = datetime.now(UTC)
            candidates: tuple[MemoryCandidate, ...] = ()
            if gate.should_retrieve and self.retriever is not None and self.retrieval_limit:
                retrieved = await self.retriever.retrieve(
                    RetrievalInput(
                        query=text,
                        history=history,
                        state=replace(state, carried_thought=None),
                        limit=self.retrieval_limit,
                        now=now,
                    )
                )
                candidates = self._validate_retrieval_batch(retrieved)
            memories = rank_candidates(
                candidates,
                limit=self.retrieval_limit,
                now=now,
                mood_valence=state.mood.valence,
            )

            appraisal = self.appraisal_policy(
                AppraisalPolicyInput(
                    state=state,
                    intent=gate.intent,
                    message=text,
                    expectation=state.expectation,
                )
            )
            if not isinstance(appraisal, AppraisalResult):
                raise TypeError("appraisal policies must return AppraisalResult")
            evolved = appraisal.state
            turn_shape = turn_shape_directive(
                evolved.mood,
                intent=gate.intent,
                history=history,
                emotions=appraisal.active_emotions,
                policy=self.turn_shape_policy,
            )
            system_prompt = build_system_prompt(
                self.base_prompt,
                evolved,
                PromptInputs(
                    emotions=appraisal.active_emotions,
                    turn_shape=turn_shape,
                ),
                policy=self.prompt_policy,
            )
            if len(system_prompt) > self.limits.max_prompt_chars:
                raise ValueError(f"system prompt exceeds {self.limits.max_prompt_chars} characters")
            request = GenerateInput(
                message=text,
                system_prompt=system_prompt,
                history=history,
                state=replace(evolved, carried_thought=None),
                intent=gate.intent,
                emotions=appraisal.active_emotions,
                decoding=decoding_params(evolved.mood),
                untrusted_context=build_untrusted_context(
                    evolved,
                    memories,
                    surface_carried_thought=not history,
                ),
            )
            chunks: list[str] = []
            output_chars = 0
            async for chunk in self.model.generate(request):
                if not isinstance(chunk, str):
                    raise TypeError("model adapters must yield strings")
                next_output_chars = output_chars + len(chunk)
                if next_output_chars > self.limits.max_output_chars:
                    raise ValueError(
                        f"model output exceeds {self.limits.max_output_chars} characters"
                    )
                chunks.append(chunk)
                output_chars = next_output_chars
                if on_token is not None:
                    on_token(chunk)
            response = "".join(chunks)
            await transaction.commit(
                state=evolved,
                messages=(user_message, Message("assistant", response)),
            )
            return TurnResult(
                text=response,
                intent=gate.intent,
                silent=False,
                mood=evolved.mood,
                emotions=appraisal.active_emotions,
                memories=memories,
            )

    def _validate_history(self, history: tuple[Message, ...]) -> None:
        if len(history) > self.limits.max_history_messages:
            raise ValueError(f"history exceeds {self.limits.max_history_messages} messages")
        history_chars = 0
        for message in history:
            history_chars += len(message.content)
            if history_chars > self.limits.max_history_chars:
                raise ValueError(f"history exceeds {self.limits.max_history_chars} characters")

    async def _gate(
        self,
        text: str,
        history: tuple[Message, ...],
        state: CompanionState,
    ) -> GateResult:
        try:
            raw = await self.model.gate(
                GateInput(
                    message=text,
                    history=history,
                    state=replace(state, carried_thought=None),
                )
            )
            if not isinstance(raw, GateResult):
                raise TypeError("model gate must return GateResult")
            return replace(
                raw,
                intent=normalize_intent(raw.intent, custom_intents=self.custom_intents),
            )
        except Exception:
            if self.gate_error_mode == "raise":
                raise
            return DEFAULT_GATE_RESULT if self.gate_error_mode == "respond" else SILENT_GATE_RESULT

    def _validate_retrieval_batch(
        self,
        candidates: Sequence[MemoryCandidate],
    ) -> tuple[MemoryCandidate, ...]:
        if not isinstance(candidates, Sequence):
            raise TypeError("memory retrievers must return a sequence")
        if len(candidates) > self.max_retrieval_candidates:
            raise ValueError(
                f"too many memory candidates: {len(candidates)} exceeds "
                f"{self.max_retrieval_candidates}"
            )
        if not all(isinstance(candidate, MemoryCandidate) for candidate in candidates):
            raise TypeError("memory retrievers must return only MemoryCandidate values")
        return tuple(candidates)

    async def presence(self, cognition: CognitionState | None = None) -> PresenceVector:
        """Load current state and return its deterministic presence vector."""
        async with self.store.transaction(self.conversation_id) as transaction:
            state = await transaction.load_state() or self.state_factory()
            return build_presence_vector(state, cognition)


__all__ = [
    "BUILTIN_INTENTS",
    "DEFAULT_ENGINE_LIMITS",
    "DEFAULT_GATE_RESULT",
    "SILENT_GATE_RESULT",
    "CompanionEngine",
    "EngineLimits",
    "GateErrorMode",
    "normalize_intent",
]

File: typescript/src/engine.ts
import { buildTurnShapeDirective, decodingParams } from './affect.js';
import type { TurnShapePolicy } from './affect.js';
import { appraiseTurn, stageInt } from './appraisal.js';
import type { AppraiseTurnInput, AppraiseTurnResult } from './appraisal.js';
import type {
  AdapterControl,
  AbortSignalLike,
  AffectState,
  DeepReadonly,
  GateResult,
  Intent,
  MemoryCandidate,
  MemoryRetriever,
  Message,
  ModelAdapter,
  RankedMemory,
  StateStore,
  StateTransaction,
  TurnCallbacks,
  TurnResult,
} from './contracts.js';
import { createAffectState, INTENTS } from './contracts.js';
import { readonlySnapshot } from './internal/snapshot.js';
import { stripPyWhitespace } from './internal/whitespace.js';
import { buildUntrustedContext, composePrompt } from './prompt.js';
import type { PromptPolicy } from './prompt.js';
import { rankCandidates } from './retrieval.js';
import { DEFAULT_PRESENCE_LABELS, buildPresenceVector, type PresenceLabels } from './surfacing.js';
import type { CognitionState, PresenceVector } from './surfacing.js';

const INTENT_SET: ReadonlySet<string> = new Set(INTENTS);
const CUSTOM_INTENT = /^[A-Z][A-Z0-9_]{0,63}$/u;

export interface EngineLimits {
  readonly maxMessageChars: number;
  readonly maxHistoryMessages: number;
  readonly maxHistoryChars: number;
  readonly maxCandidates: number;
  readonly maxMemories: number;
  readonly maxMemoryChars: number;
  readonly maxPromptChars: number;
  readonly maxOutputChars: number;
  /** Maximum turns admitted by one engine instance, including the active turn. */
  readonly maxQueuedTurns: number;
}

export const DEFAULT_ENGINE_LIMITS: Readonly<EngineLimits> = Object.freeze({
  maxMessageChars: 16_000,
  maxHistoryMessages: 200,
  maxHistoryChars: 128_000,
  maxCandidates: 100,
  maxMemories: 20,
  maxMemoryChars: 8_000,
  maxPromptChars: 128_000,
  maxOutputChars: 32_000,
  maxQueuedTurns: 32,
});

export type AppraisalPolicy = (
  input: Readonly<AppraiseTurnInput>,
) => AppraiseTurnResult;

/** Current conversational OCC/PAD mapping, exposed as the reference policy. */
export const DEFAULT_APPRAISAL_POLICY: AppraisalPolicy = (input) => appraiseTurn(input);

export interface AffectEngineOptions {
  readonly model: ModelAdapter;
  readonly store: StateStore;
  readonly retriever?: MemoryRetriever;
  readonly instruction?: string;
  readonly promptPolicy?: PromptPolicy;
  readonly turnShapePolicy?: TurnShapePolicy;
  /** Wording of the presence surface; defaults to the conversational phrasing. */
  readonly presenceLabels?: PresenceLabels;
  readonly stateFactory?: () => AffectState;
  readonly retrievalLimit?: number;
  /** Opt-in fallback; omitted means gate errors and malformed output propagate. */
  readonly gateFallback?: GateResult;
  /** Additional normalized labels; custom labels intentionally have no built-in appraisal impulse. */
  readonly customIntents?: Iterable<string>;
  readonly appraisalPolicy?: AppraisalPolicy;
  readonly limits?: Partial<EngineLimits>;
  readonly now?: () => Date;
}

function parseIntent(value: unknown, allowed: ReadonlySet<string>): Intent {
  if (typeof value !== 'string') throw new TypeError('gate intent must be a string');
  const normalized = value.trim().toUpperCase();
  if (!allowed.has(normalized)) throw new RangeError(`gate intent is not supported: ${value}`);
  return normalized as Intent;
}

function parseGateResult(value: unknown, allowed: ReadonlySet<string>): GateResult {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new TypeError('gate output must be an object');
  }
  const candidate = value as Record<string, unknown>;
  if (typeof candidate.shouldRespond !== 'boolean') {
    throw new TypeError('gate shouldRespond must be a boolean');
  }
  if (typeof candidate.shouldRetrieve !== 'boolean') {
    throw new TypeError('gate shouldRetrieve must be a boolean');
  }
  return Object.freeze({
    intent: parseIntent(candidate.intent, allowed),
    shouldRespond: candidate.shouldRespond,
    shouldRetrieve: candidate.shouldRetrieve,
  });
}

function resolveIntents(customIntents: Iterable<string> | undefined): ReadonlySet<string> {
  const allowed = new Set(INTENT_SET);
  if (customIntents === undefined) return allowed;
  if (typeof customIntents === 'string' || customIntents === null
    || typeof customIntents[Symbol.iterator] !== 'function') {
    throw new TypeError('customIntents must be an iterable of strings');
  }
  for (const value of customIntents) {
    if (typeof value !== 'string') throw new TypeError('custom intents must be strings');
    const normalized = value.trim().toUpperCase();
    if (!CUSTOM_INTENT.test(normalized)) {
      throw new RangeError('custom intent must match [A-Z][A-Z0-9_]{0,63}');
    }
    allowed.add(normalized);
  }
  return allowed;
}

function resolveLimits(input: Partial<EngineLimits> | undefined): Readonly<EngineLimits> {
  const limits = { ...DEFAULT_ENGINE_LIMITS, ...(input ?? {}) };
  for (const [name, value] of Object.entries(limits)) {
    if (!Number.isSafeInteger(value) || value < 0) {
      throw new RangeError(`${name} must be a non-negative safe integer`);
    }
  }
  if (limits.maxMessageChars === 0
    || limits.maxHistoryMessages === 0
    || limits.maxHistoryChars === 0
    || limits.maxPromptChars === 0
    || limits.maxOutputChars === 0
    || limits.maxQueuedTurns === 0) {
    throw new RangeError('message, history, prompt, output, and queued-turn limits must be positive');
  }
  return Object.freeze(limits);
}

function validateHistory(
  value: unknown,
  limits: Readonly<EngineLimits>,
): asserts value is ReadonlyArray<DeepReadonly<Message>> {
  if (!Array.isArray(value)) throw new TypeError('store history must be an array');
  if (value.length > limits.maxHistoryMessages) {
    throw new RangeError(`history exceeds ${limits.maxHistoryMessages} messages`);
  }
  let characters = 0;
  for (const [index, message] of value.entries()) {
    if (message === null || typeof message !== 'object' || Array.isArray(message)) {
      throw new TypeError(`history message ${index} must be an object`);
    }
    const candidate = message as Record<string, unknown>;
    if (candidate.role !== 'user' && candidate.role !== 'assistant') {
      throw new RangeError(`history message ${index} has an unsupported role`);
    }
    if (typeof candidate.content !== 'string') {
      throw new TypeError(`history message ${index} content must be a string`);
    }
    characters += candidate.content.length;
    if (characters > limits.maxHistoryChars) {
      throw new RangeError(`history exceeds ${limits.maxHistoryChars} characters`);
    }
  }
}

function validateControl(control: AdapterControl): void {
  const signal = control.signal;
  if (signal !== undefined
    && (signal === null
      || typeof signal !== 'object'
      || typeof signal.aborted !== 'boolean'
      || typeof signal.addEventListener !== 'function'
      || typeof signal.removeEventListener !== 'function')) {
    throw new TypeError('signal must be AbortSignal-compatible');
  }
  if (control.deadline !== undefined
    && (!(control.deadline instanceof Date) || !Number.isFinite(control.deadline.getTime()))) {
    throw new TypeError('deadline must be a valid Date');
  }
}

function abortError(signal: AbortSignalLike): Error {
  const error = new Error('turn aborted', { cause: signal.reason });
  error.name = 'AbortError';
  return error;
}

function timeoutError(): Error {
  const error = new Error('turn deadline exceeded');
  error.name = 'TimeoutError';
  return error;
}

function stopIfNeeded(control: AdapterControl): void {
  if (control.signal?.aborted) throw abortError(control.signal);
  if (control.deadline !== undefined && Date.now() >= control.deadline.getTime()) {
    throw timeoutError();
  }
}

function isControlError(value: unknown): boolean {
  if (value === null || typeof value !== 'object') return false;
  const name = (value as { readonly name?: unknown }).name;
  return name === 'AbortError' || name === 'TimeoutError';
}

function snapshotCallbacks(callbacks: TurnCallbacks): Readonly<TurnCallbacks> {
  if (callbacks === null || typeof callbacks !== 'object') {
    throw new TypeError('callbacks must be an object');
  }
  const signal = callbacks.signal;
  const deadline = callbacks.deadline;
  const onToken = callbacks.onToken;
  const rawControl: AdapterControl = {
    ...(signal === undefined ? {} : { signal }),
    ...(deadline === undefined ? {} : { deadline }),
  };
  validateControl(rawControl);
  if (onToken !== undefined && typeof onToken !== 'function') {
    throw new TypeError('onToken must be a function');
  }
  const snapshot: TurnCallbacks = {
    ...rawControl,
    ...(deadline === undefined ? {} : { deadline: new Date(deadline.getTime()) }),
    ...(onToken === undefined ? {} : { onToken }),
  };
  return Object.freeze(snapshot);
}

const MAX_TIMER_DELAY_MS = 2_147_483_647;
const RUNTIME_TIMERS = globalThis as unknown as {
  setTimeout(handler: () => void, delay: number): unknown;
  clearTimeout(handle: unknown): void;
};

function waitWithControl<Result>(
  work: PromiseLike<Result>,
  control: AdapterControl,
): Promise<Result> {
  try {
    stopIfNeeded(control);
  } catch (error) {
    return Promise.reject(error);
  }

  return new Promise<Result>((resolve, reject) => {
    let finished = false;
    let timer: unknown;
    let listening = false;
    const signal = control.signal;

    const cleanup = (): void => {
      if (timer !== undefined) RUNTIME_TIMERS.clearTimeout(timer);
      if (listening && signal !== undefined) {
        try {
          signal.removeEventListener('abort', onAbort);
        } catch {
          // A broken signal must not mask the operation's settled result.
        }
      }
    };
    const settle = (operation: () => void): void => {
      if (finished) return;
      finished = true;
      cleanup();
      operation();
    };
    const resolveOnce = (value: Result): void => { settle(() => resolve(value)); };
    const rejectOnce = (error: unknown): void => { settle(() => reject(error)); };
    const onAbort = (): void => {
      if (signal !== undefined) rejectOnce(abortError(signal));
    };
    const scheduleDeadline = (): void => {
      const deadline = control.deadline;
      if (deadline === undefined || finished) return;
      const remaining = deadline.getTime() - Date.now();
      if (remaining <= 0) {
        rejectOnce(timeoutError());
        return;
      }
      timer = RUNTIME_TIMERS.setTimeout(
        scheduleDeadline,
        Math.min(remaining, MAX_TIMER_DELAY_MS),
      );
    };

    Promise.resolve(work).then(resolveOnce, rejectOnce);
    try {
      if (signal !== undefined) {
        listening = true;
        signal.addEventListener('abort', onAbort, { once: true });
        if (signal.aborted) onAbort();
      }
      scheduleDeadline();
    } catch (error) {
      rejectOnce(error);
    }
  });
}

function closeIterator(iterator: AsyncIterator<unknown>): void {
  try {
    const completion = iterator.return?.();
    if (completion !== undefined) void Promise.resolve(completion).catch(() => undefined);
  } catch {
    // Iterator cleanup is best-effort and must not mask the turn failure.
  }
}

function adapterControl(control: AdapterControl): AdapterControl {
  return {
    ...(control.signal === undefined ? {} : { signal: control.signal }),
    ...(control.deadline === undefined ? {} : { deadline: new Date(control.deadline.getTime()) }),
  };
}

function adapterInput<Input extends object>(input: Input, control: AdapterControl): Readonly<Input & AdapterControl> {
  const snapshot = readonlySnapshot(input);
  return Object.freeze({ ...snapshot, ...adapterControl(control) }) as Readonly<Input & AdapterControl>;
}

function parseCandidates(value: unknown, limits: Readonly<EngineLimits>): MemoryCandidate[] {
  if (!Array.isArray(value)) throw new TypeError('memory retriever must return an array');
  if (value.length > limits.maxCandidates) {
    throw new RangeError(`memory retriever returned more than ${limits.maxCandidates} candidates`);
  }
  let textChars = 0;
  for (const [index, candidate] of value.entries()) {
    if (candidate === null || typeof candidate !== 'object' || Array.isArray(candidate)) {
      throw new TypeError(`memory candidate ${index} must be an object`);
    }
    const text = (candidate as Record<string, unknown>).text;
    if (typeof text !== 'string') throw new TypeError(`memory candidate ${index} text must be a string`);
    textChars += text.length;
    if (textChars > limits.maxMemoryChars) {
      throw new RangeError(`memory evidence exceeds ${limits.maxMemoryChars} characters`);
    }
  }
  return value as MemoryCandidate[];
}

function parseAppraisalResult(value: unknown): AppraiseTurnResult {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new TypeError('appraisal policy must return an AppraiseTurnResult object');
  }
  const candidate = value as Record<string, unknown>;
  if (candidate.mood === undefined
    || candidate.baselineValence === undefined
    || candidate.occCarry === undefined) {
    throw new TypeError('appraisal policy result is missing state fields');
  }
  const validated = createAffectState({
    mood: candidate.mood as NonNullable<AffectState['mood']>,
    baselineValence: candidate.baselineValence as number,
    occCarry: candidate.occCarry as Readonly<Record<string, number>>,
  });
  const emotions = candidate.activeEmotions;
  if (emotions === null || typeof emotions !== 'object' || Array.isArray(emotions)) {
    throw new TypeError('appraisal policy activeEmotions must be an object');
  }
  const activeEmotions: Record<string, number> = {};
  for (const [name, intensity] of Object.entries(emotions)) {
    if (!name) throw new RangeError('appraisal emotion names must not be empty');
    if (typeof intensity !== 'number' || !Number.isFinite(intensity)
      || intensity < 0 || intensity > 1) {
      throw new RangeError(`appraisal emotion ${name} must be finite and between 0 and 1`);
    }
    activeEmotions[name] = intensity;
  }
  return {
    mood: validated.mood,
    activeEmotions,
    occCarry: { ...validated.occCarry },
    baselineValence: validated.baselineValence,
  };
}

/** Injected gate → retrieval → appraisal → prompt → generation pipeline. */
export class AffectEngine {
  private readonly model: ModelAdapter;
  private readonly store: StateStore;
  private readonly retriever: MemoryRetriever | undefined;
  private readonly instruction: string;
  private readonly promptPolicy: PromptPolicy | undefined;
  private readonly turnShapePolicy: TurnShapePolicy | undefined;
  private readonly presenceLabels: PresenceLabels;
  private readonly stateFactory: (() => AffectState) | undefined;
  private readonly retrievalLimit: number;
  private readonly now: (() => Date) | undefined;
  private readonly limits: Readonly<EngineLimits>;
  private readonly gateFallback: GateResult | undefined;
  private readonly allowedIntents: ReadonlySet<string>;
  private readonly appraisalPolicy: AppraisalPolicy;
  private tail: Promise<void> = Promise.resolve();
  private admittedTurns = 0;

  constructor(options: AffectEngineOptions) {
    if (options === null || typeof options !== 'object') {
      throw new TypeError('options must be an object');
    }
    const {
      model,
      store,
      retriever,
      instruction = '',
      promptPolicy,
      turnShapePolicy,
      presenceLabels,
      stateFactory,
      retrievalLimit: limit = 6,
      gateFallback,
      customIntents,
      appraisalPolicy,
      limits,
      now,
    } = options;
    if (!Number.isSafeInteger(limit) || limit < 0) {
      throw new RangeError('retrievalLimit must be a non-negative safe integer');
    }
    if (typeof instruction !== 'string') {
      throw new TypeError('instruction must be a string');
    }
    this.limits = resolveLimits(limits);
    if (this.limits.maxCandidates > 0 && limit > this.limits.maxCandidates) {
      throw new RangeError('retrievalLimit must not exceed limits.maxCandidates');
    }
    this.allowedIntents = resolveIntents(customIntents);
    if (appraisalPolicy !== undefined && typeof appraisalPolicy !== 'function') {
      throw new TypeError('appraisalPolicy must be a function');
    }
    if (stateFactory !== undefined && typeof stateFactory !== 'function') {
      throw new TypeError('stateFactory must be a function');
    }
    if (now !== undefined && typeof now !== 'function') {
      throw new TypeError('now must be a function');
    }
    this.model = model;
    this.store = store;
    this.retriever = retriever;
    this.instruction = instruction;
    this.promptPolicy = promptPolicy === undefined
      ? undefined
      : readonlySnapshot(promptPolicy) as PromptPolicy;
    this.turnShapePolicy = turnShapePolicy === undefined
      ? undefined
      : readonlySnapshot(turnShapePolicy) as TurnShapePolicy;
    this.presenceLabels = presenceLabels === undefined
      ? DEFAULT_PRESENCE_LABELS
      : readonlySnapshot(presenceLabels) as PresenceLabels;
    this.stateFactory = stateFactory;
    this.retrievalLimit = limit;
    this.now = now;
    this.appraisalPolicy = appraisalPolicy ?? DEFAULT_APPRAISAL_POLICY;
    this.gateFallback = gateFallback === undefined
      ? undefined
      : parseGateResult(gateFallback, this.allowedIntents);
  }

  turn(message: string, callbacks: TurnCallbacks = {}): Promise<TurnResult> {
    if (typeof message !== 'string') return Promise.reject(new TypeError('message must be a string'));
    if (!stripPyWhitespace(message)) {
      return Promise.reject(new RangeError('message must not be empty'));
    }
    if (message.length > this.limits.maxMessageChars) {
      return Promise.reject(new RangeError(`message exceeds ${this.limits.maxMessageChars} characters`));
    }
    let control: Readonly<TurnCallbacks>;
    try {
      control = snapshotCallbacks(callbacks);
      stopIfNeeded(control);
    } catch (error) {
      return Promise.reject(error);
    }
    if (this.admittedTurns >= this.limits.maxQueuedTurns) {
      return Promise.reject(new RangeError('maximum queued turns exceeded'));
    }

    this.admittedTurns += 1;
    const task = this.tail.then(() => this.store.transaction(
      (transaction) => this.runTurn(transaction, message, control),
    ));
    this.tail = task.then(() => undefined, () => undefined);
    return task.finally(() => { this.admittedTurns -= 1; });
  }

  async presence(cognition: CognitionState = {}): Promise<PresenceVector> {
    const stored = await this.store.loadState();
    const state = createAffectState(stored ?? this.stateFactory?.() ?? {});
    return readonlySnapshot(
      buildPresenceVector(state, cognition, this.presenceLabels),
    ) as PresenceVector;
  }

  private currentTime(): Date {
    const now = this.now?.() ?? new Date();
    if (!(now instanceof Date) || !Number.isFinite(now.getTime())) {
      throw new TypeError('now must return a valid Date');
    }
    return new Date(now.getTime());
  }

  private async runTurn(
    transaction: StateTransaction,
    message: string,
    callbacks: TurnCallbacks,
  ): Promise<TurnResult> {
    stopIfNeeded(callbacks);
    const history = readonlySnapshot(await transaction.listMessages());
    validateHistory(history, this.limits);
    const stored = await transaction.loadState();
    const state = readonlySnapshot(createAffectState(
      stored ?? this.stateFactory?.() ?? {},
    ));
    const adapterState = readonlySnapshot(createAffectState({ ...state, carriedThought: null }));
    stopIfNeeded(callbacks);

    let gate: GateResult;
    let gateFallback = false;
    try {
      const rawGate = await waitWithControl(this.model.gate(adapterInput({
        message,
        history,
        state: adapterState,
      }, callbacks)), callbacks);
      gate = parseGateResult(rawGate, this.allowedIntents);
    } catch (error) {
      if (isControlError(error)) throw error;
      stopIfNeeded(callbacks);
      if (this.gateFallback === undefined) throw error;
      gate = this.gateFallback;
      gateFallback = true;
    }

    await transaction.appendMessage({ role: 'user', content: message });
    if (!gate.shouldRespond) {
      return readonlySnapshot({
        status: 'silent' as const,
        text: null,
        intent: gate.intent,
        mood: state.mood,
        emotions: {},
        memories: [],
        decoding: null,
        gateFallback,
      }) as TurnResult;
    }

    const now = this.currentTime();
    let candidates: MemoryCandidate[] = [];
    const retrievalLimit = this.retrievalLimit;
    if (gate.shouldRetrieve
      && this.retriever
      && retrievalLimit > 0
      && this.limits.maxCandidates > 0
      && this.limits.maxMemories > 0
      && this.limits.maxMemoryChars > 0) {
      const rawCandidates = await waitWithControl(this.retriever.retrieve(adapterInput({
        query: message,
        history,
        state: adapterState,
        limit: retrievalLimit,
        now: new Date(now.getTime()),
      }, callbacks)), callbacks);
      candidates = parseCandidates(rawCandidates, this.limits);
    }
    const memories: RankedMemory[] = rankCandidates(candidates, {
      limit: Math.min(retrievalLimit, this.limits.maxMemories),
      now,
      moodValence: state.mood.valence,
    });

    const appraisal = readonlySnapshot(parseAppraisalResult(this.appraisalPolicy(readonlySnapshot({
      mood: state.mood,
      personality: state.personality,
      goals: state.goals,
      stageInt: stageInt(state.relationship.stage),
      baselineValence: state.baselineValence,
      attachmentLonging: state.attachment.longing,
      intent: gate.intent,
      occCarry: state.occCarry,
      expectation: state.expectation,
      message,
    }))));
    const evolved = readonlySnapshot(createAffectState({
      ...state,
      mood: appraisal.mood,
      baselineValence: appraisal.baselineValence,
      occCarry: appraisal.occCarry,
    }));
    await transaction.saveState(evolved);

    const decoding = decodingParams(evolved.mood);
    const turnDirective = buildTurnShapeDirective(
      {
        mood: evolved.mood,
        intent: gate.intent,
        history,
        emotions: appraisal.activeEmotions,
      },
      this.turnShapePolicy,
    );
    const promptContext = {
      instruction: this.instruction,
      state: { mood: evolved.mood },
      emotions: appraisal.activeEmotions,
      decoding,
      turnDirective,
    };
    const systemPrompt = this.promptPolicy
      ? composePrompt(promptContext, this.promptPolicy)
      : composePrompt(promptContext);
    if (systemPrompt.length > this.limits.maxPromptChars) {
      throw new RangeError(`system prompt exceeds ${this.limits.maxPromptChars} characters`);
    }
    const untrustedContext = buildUntrustedContext(evolved, memories, {
      maxChars: this.limits.maxMemoryChars,
      surfaceCarriedThought: history.length === 0,
    });
    const generationState = readonlySnapshot(createAffectState({
      ...evolved,
      carriedThought: null,
    }));

    const chunks: string[] = [];
    let accumulated = '';
    const generation = this.model.generate(adapterInput({
      message,
      systemPrompt,
      history,
      state: generationState,
      intent: gate.intent,
      emotions: appraisal.activeEmotions,
      decoding,
      untrustedContext,
    }, callbacks));
    const iterator = generation[Symbol.asyncIterator]();
    let iteratorDone = false;
    try {
      while (true) {
        const result = await waitWithControl(iterator.next(), callbacks);
        if (result.done) {
          iteratorDone = true;
          break;
        }
        const chunk = result.value;
        if (typeof chunk !== 'string') throw new TypeError('model adapters must yield strings');
        if (accumulated.length + chunk.length > this.limits.maxOutputChars) {
          throw new RangeError(`model output exceeds ${this.limits.maxOutputChars} characters`);
        }
        chunks.push(chunk);
        accumulated += chunk;
        if (callbacks.onToken !== undefined) {
          await waitWithControl(
            Promise.resolve(callbacks.onToken(chunk, accumulated)),
            callbacks,
          );
        }
      }
    } finally {
      if (!iteratorDone) closeIterator(iterator);
    }
    const text = chunks.join('');
    await transaction.appendMessage({ role: 'assistant', content: text });
    return readonlySnapshot({
      status: 'responded' as const,
      text,
      intent: gate.intent,
      mood: evolved.mood,
      emotions: appraisal.activeEmotions,
      memories,
      decoding,
      gateFallback,
    }) as TurnResult;
  }
}
